clients/src/main/java/org/apache/kafka/common/requests/UpdateMetadataRequest.java

```diff
@@ -218,257 +116,68 @@ public String toString() {
         }
     }
 
-    public static final class PartitionState {
-        public final BasePartitionState basePartitionState;
-        public final List<Integer> offlineReplicas;
-
-        public PartitionState(int controllerEpoch,
-                              int leader,
-                              int leaderEpoch,
-                              List<Integer> isr,
-                              int zkVersion,
-                              List<Integer> replicas,
-                              List<Integer> offlineReplicas) {
-            this.basePartitionState = new BasePartitionState(controllerEpoch, leader, leaderEpoch, isr, zkVersion, replicas);
-            this.offlineReplicas = offlineReplicas;
-        }
-
-        private PartitionState(Struct struct) {
-            int controllerEpoch = struct.get(CONTROLLER_EPOCH);
-            int leader = struct.get(LEADER);
-            int leaderEpoch = struct.get(LEADER_EPOCH);
-
-            Object[] isrArray = struct.get(ISR);
-            List<Integer> isr = new ArrayList<>(isrArray.length);
-            for (Object r : isrArray)
-                isr.add((Integer) r);
+    private final UpdateMetadataRequestData data;
 
-            int zkVersion = struct.get(ZK_VERSION);
-
-            Object[] replicasArray = struct.get(REPLICAS);
-            List<Integer> replicas = new ArrayList<>(replicasArray.length);
-            for (Object r : replicasArray)
-                replicas.add((Integer) r);
-
-            this.basePartitionState = new BasePartitionState(controllerEpoch, leader, leaderEpoch, isr, zkVersion, replicas);
+    UpdateMetadataRequest(UpdateMetadataRequestData data, short version) {
+        super(ApiKeys.UPDATE_METADATA, version);
+        this.data = data;
+        // Do this from the constructor to make it thread-safe (even though it's only needed when some methods are called)
+        normalize();
+    }
 
-            this.offlineReplicas = new ArrayList<>();
-            if (struct.hasField(OFFLINE_REPLICAS)) {
-                Object[] offlineReplicasArray = struct.get(OFFLINE_REPLICAS);
-                for (Object r : offlineReplicasArray)
-                    offlineReplicas.add((Integer) r);
+    private void normalize() {
+        if (version() < 3) {
```

Comment:
it would be nice to have a short description of what was different in version 0,1,2